packages/block-library/src/embed/embed-placeholder.native.js

```diff
@@ -26,21 +37,37 @@ const EmbedPlaceholder = ( { icon, label, onFocus } ) => {
 	);
 
 	return (
-		<TouchableWithoutFeedback
-			accessibilityRole={ 'button' }
-			accessibilityHint={ __( 'Double tap to add a link.' ) }
-			onPress={ ( event ) => {
-				onFocus( event );
-			} }
-		>
-			<View style={ emptyStateContainerStyle }>
-				<View style={ styles.modalIcon }>{ icon }</View>
-				<Text style={ emptyStateTitleStyle }>{ label }</Text>
-				<Text style={ styles.emptyStateDescription }>
-					{ __( 'ADD LINK' ) }
-				</Text>
-			</View>
-		</TouchableWithoutFeedback>
+		<>
+			{ value ? (
+				<Text>{ value }</Text>
+			) : (
+				<TouchableWithoutFeedback
+					accessibilityRole={ 'button' }
+					accessibilityHint={ __( 'Double tap to add a link.' ) }
+					onPress={ ( event ) => {
+						onFocus( event );
+						setIsEmbedSheetVisible( true );
+					} }
+				>
+					<View style={ emptyStateContainerStyle }>
+						<View style={ styles.modalIcon }>{ icon }</View>
+						<Text style={ emptyStateTitleStyle }>{ label }</Text>
+						<Text style={ styles.emptyStateDescription }>
+							{ __( 'ADD LINK' ) }
+						</Text>
+					</View>
+				</TouchableWithoutFeedback>
+			) }
+			<EmbedBottomSheet
+				value={ value }
+				isVisible={ isEmbedSheetVisible }
+				onClose={ () => setIsEmbedSheetVisible( false ) }
+				onSetAttributes={ ( { url } ) => {
+					onChange( { target: { value: url } } );
+					onSubmit();
```

Comment:
`onSubmit` is the callback that actually updates the attributes of the block:
https://github.com/WordPress/gutenberg/blob/8d3db371b033d4e331702a7a837978fac79e82ca/packages/block-library/src/embed/edit.js#L204-L211

Reply:
The behavior of setting a URL differs a bit from the web version, there `onChange` is used to update the internal state when the user types the URL, in the native version this is done within the `EmbedBottomSheet` component so for update attributes we have to explicitly call both callbacks:
- `onChange` with the URL to update the internal state
- `onSubmit` to set the attributes with the previous updated internal state

Reply:
Speaking of the the `onSubmit()`, looks like it's called twice when dismissing the bottom sheet. Not sure if that's what you also witness @fluiddot?

Reply:
Oh that's interesting, I didn't spot it when testing, I'll review this part deeper, thanks for the heads-up 🙇 !

Reply:
True, the `onSubmit` callback gets called twice, looks like the `onCloseBottomSheet` callback from the bottom sheet component is called twice from different places and end up setting the attributes two times:
1. `onBackdropPress` prop from the Modal:  https://github.com/WordPress/gutenberg/blob/779fdaa94263611a3ccb5f7fdd5f312b43e5cbea/packages/components/src/mobile/bottom-sheet/index.native.js#L496
2. `onDismiss` callback from Modal: https://github.com/WordPress/gutenberg/blob/779fdaa94263611a3ccb5f7fdd5f312b43e5cbea/packages/components/src/mobile/bottom-sheet/index.native.js#L316

When `onCloseBottomSheet` is called, internally it sets the attributes by the following flow:
- The `LinkSettings` components sets the `handleClosingBottomSheet` callback:  https://github.com/WordPress/gutenberg/blob/779fdaa94263611a3ccb5f7fdd5f312b43e5cbea/packages/components/src/mobile/link-settings/index.native.js#L101
- `onCloseBottomSheet` calls the callback `handleClosingBottomSheet` (this callback is the one previously set in the above line):
https://github.com/WordPress/gutenberg/blob/779fdaa94263611a3ccb5f7fdd5f312b43e5cbea/packages/components/src/mobile/bottom-sheet/index.native.js#L343
- The callback internally calls the function to set the attributes:
https://github.com/WordPress/gutenberg/blob/779fdaa94263611a3ccb5f7fdd5f312b43e5cbea/packages/components/src/mobile/link-settings/index.native.js#L177

At this point, I'm leaning towards using the `onDismiss` callback to trigger the submit as it looks the only one that would be called only once and use `onSetAttributes` to update the internal state as the web version does 🤔 .

Reply:
Ah, thanks for confirming Carlos! the "duplicate call" of `onSubmit` is worrying, but not sure yet if we should block this PR on it. I'll check develop.

Reply:
After [this commit](https://github.com/WordPress/gutenberg/pull/31343/commits/0154c6f0131b633e97ca79dc9a25791a3fffd6ac), now the `onSubmit` callback is not called twice because it's located in a different callback, before it was triggered in `onSetAttributes` and now is [`onDismiss`](https://github.com/WordPress/gutenberg/pull/31343/files#diff-7d155d7b9eb26681fd43d964649784c4799d33b187271b3288e604f9c9728b8fR39), so this is no longer an issue.

`onSetAttributes ` callback will still be called twice, as it's described in the previous comment, but I think it shouldn't produce any bad side effect as it's only updating the internal state of the embed sheet component ([code reference](https://github.com/WordPress/gutenberg/pull/31343/files#diff-7d155d7b9eb26681fd43d964649784c4799d33b187271b3288e604f9c9728b8fR48-R50)).

Reply:
> onSetAttributes callback will still be called twice, as it's described in the previous comment, but I think it shouldn't produce any bad side effect as it's only updating the internal state of the embed sheet component ([code reference](https://github.com/WordPress/gutenberg/pull/31343/files#diff-7d155d7b9eb26681fd43d964649784c4799d33b187271b3288e604f9c9728b8fR48-R50)).

I think it will also call the [block's onChange](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-editor/src/components/block-list/block.native.js#L407) and update the attributes in the store (makes sense), maybe causing some extra renders. I'm not fully sure. It still doesn't sound like bad side effects, except perhaps for performance. Anyway, I think this is not a blocker for this PR. 👍

Reply:
> I think it will also call the block's onChange and update the attributes in the store (makes sense), maybe causing some extra renders. I'm not fully sure. It still doesn't sound like bad side effects, except perhaps for performance.

I'd like to note that the code displayed above is outdated, the current code only does the internal update when the `setAttributes` is triggered, so it shouldn't call [block's onChange](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-editor/src/components/block-list/block.native.js#L407):
https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-library/src/embed/embed-bottom-sheet.native.js#L48-L62

I also checked the code of the component that handles the link settings ([`LinkSettings`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/link-settings/index.native.js)) just in case it's updating the attributes internally but it doesn't, it only triggers the callback. Here is the code in charge of notifying the update of the attributes:
https://github.com/WordPress/gutenberg/blob/326ebc6044418d91df2d58c0c9dce7548e031ab1/packages/components/src/mobile/link-settings/index.native.js#L161-L174

Reply:
>  so it shouldn't call [block's onChange](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-editor/src/components/block-list/block.native.js#L407):

Hmm, I do see the line hit in the debugger on 32c4466f62 🤔

Reply:
> Hmm, I do see the line getting hit in the debugger on 32c4466 🤔

Oh I think I wasn't clear on my explanation, the `setAttributes` callback will be called twice but it won't trigger the [block's onChange](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-editor/src/components/block-list/block.native.js#L407), instead, it will be triggered by the `onSubmit` callback when the modal is dismissed not when the [`LinkSettings`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/link-settings/index.native.js) component notifies the update of the attributes.

Here I added a patch diff that logs in the console calls to both callbacks `setAttributes` and [block's onChange](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-editor/src/components/block-list/block.native.js#L407) in case it makes my explanation clearer 😄 :
```diff
diff --git a/packages/block-library/src/embed/embed-bottom-sheet.native.js b/packages/block-library/src/embed/embed-bottom-sheet.native.js
index e7e69fb994302582a9fa8fb22f10bfd64f5b9156..4e910047ea6fa97deedf591057f3c4d9f875b470 100644
--- a/packages/block-library/src/embed/embed-bottom-sheet.native.js
+++ b/packages/block-library/src/embed/embed-bottom-sheet.native.js
@@ -46,6 +46,7 @@ const EmbedBottomSheet = ( { value, isVisible, onClose, onSubmit } ) => {
 	}, [ url, onSubmit ] );
 
 	function setAttributes( attributes ) {
+		console.log( 'setAttributes', attributes );
 		setURL( attributes.url );
 	}
 
diff --git a/packages/block-editor/src/components/block-list/block.native.js b/packages/block-editor/src/components/block-list/block.native.js
index 018cf4760a07ebc6f208452b9ed01c6de0d0e57a..418cb025b260d7311284689967f3661a004e2585 100644
--- a/packages/block-editor/src/components/block-list/block.native.js
+++ b/packages/block-editor/src/components/block-list/block.native.js
@@ -404,6 +404,7 @@ export default compose( [
 				selectBlock( clientId, initialPosition );
 			},
 			onChange: ( attributes ) => {
+				console.log( 'onChange', attributes );
 				updateBlockAttributes( ownProps.clientId, attributes );
 			},
 			onReplace( blocks, indexToSelect ) {
```

As you can see `setAttributes` is called two times but `onChange` only once, so the whole block shouldn't get re-rendered, however, it will the `EmbedBottomSheet` component as its internal state is updated:

<img width="573" alt="Screenshot 2021-07-01 at 10 29 41" src="https://user-images.githubusercontent.com/14905380/124092537-420b9f80-da57-11eb-99aa-592c2baef4da.png">